sycl/include/sycl/ext/intel/esimd/memory.hpp

```diff
@@ -14302,6 +14302,254 @@ named_barrier_signal(uint8_t barrier_id, uint8_t producer_consumer_mode,
+    Result.iupdate(Indices, src, mask);
+    return Result;
+  } else {
+    simd<Tx, M> Result;
+    simd<uint16_t, M> Indices(0);
+    Indices.template select<N, 1>(0) = global_offset + offset;
+    simd_mask<M> Newmask(0);
+    Newmask.template select<N, 1>(0) = mask;
+    Result.iupdate(Indices, src, Newmask);
+    return Result.template select<N, 1>(0);
+  }
+}
+
+/// template <typename Tx, int N,
+///       typename PropertyListT = oneapi::experimental::empty_properties_t>
+/// simd<Tx, N>
+/// mask_compress_load(Tx *p, simd_mask<N> mask, PropertyListT props = {});
+/// Mask compressed load from USM memory location.
+/// The function reads data from a memory location using following algorithm:
+///
+/// \code{.cpp}
+///
+/// int Index = 0;
+/// for (int i = 0; i < N; ++i) {
+///       if (Mask[i]) {
+///           Result[i] = *(p + Index);
+///           ++Index;
+///       }
+/// }
+/// \endcode
+///
+///
+/// @tparam Tx is element type.
+/// @tparam N is the data size
+/// @param p is the base address for this operation.
+/// @param mask is the mask determining which elements will be read
+/// @param props The compile-time properties. Only cache hint
+/// properties are used.
+///
+template <typename Tx, int N,
+          typename PropertyListT = oneapi::experimental::empty_properties_t>
+__ESIMD_API std::enable_if_t<
+    ext::oneapi::experimental::is_property_list_v<PropertyListT>, simd<Tx, N>>
+mask_compress_load(Tx *p, simd_mask<N> mask, PropertyListT props = {}) {
+  simd<Tx, N> Src = block_load<Tx, N>(p, props);
+  return mask_compress_load(Src, 0, mask);
+}
+
+/// template <typename Tx, int N, typename AccessorTy,
+///       typename PropertyListT = oneapi::experimental::empty_properties_t>
+/// simd<Tx, N>
+/// mask_compress_load(AccessorTy acc, simd_mask<N> mask, PropertyListT props =
+/// {});
+/// Mask compressed load from accessor memory (could be local or device
+/// accessor). The function reads data from a memory location using following
+/// algorithm:
+///
+/// \code{.cpp}
+///
+/// int Index = 0;
+/// for (int i = 0; i < N; ++i) {
+///       if (Mask[i])
+///           Result[i] = acc[global_offset + Index++];
+/// }
+/// \endcode
+///
+///
+/// @tparam Tx is element type.
+/// @tparam N is the data size
+/// @param acc is the accessor to read from.
+/// @param global_offset is the global offset in bytes
+/// @param mask is the mask determining which elements will be read
+/// @param props The compile-time properties. Only cache hint
+/// properties are used.
+///
+template <typename Tx, int N, typename AccessorTy,
+          typename PropertyListT = oneapi::experimental::empty_properties_t>
+__ESIMD_API std::enable_if_t<
+    ext::oneapi::experimental::is_property_list_v<PropertyListT> &&
+        detail::is_accessor_with_v<AccessorTy,
+                                   detail::accessor_mode_cap::can_read>,
+    simd<Tx, N>>
+mask_compress_load(AccessorTy acc, uint32_t global_offset, simd_mask<N> mask,
```

Comment:
AFAIR, both LLVM and CPU hardware support [expand load](https://llvm.org/docs/LangRef.html#llvm-masked-expandload-intrinsics) operations, that should be emulated by conditional gather. What is the idea of "compress load"?